draft-irtf-cfrg-opaque.md

```diff
@@ -967,93 +1102,273 @@ enc_server_info
+12. Create KE2 ke2 with (ike2, enc_server_info, server_mac)
+13. Output (ke2, ke1.client_info, expected_client_mac, session_key)
 ~~~
-server_mac_key =
-  Expand-Label(handshake_secret, "server mac", "", Nx)
-client_mac_key =
-  Expand-Label(handshake_secret, "client mac", "", Nx)
-handshake_encrypt_key =
-  Expand-Label(handshake_secret, "handshake enc", "", Nx)
+
 ~~~
+AKEClientFinalize(client_secret, client_identity, client_private_key, server_identity, server_public_key, ke1, ke2)
+
+Parameters:
+- Nn, the nonce length
 
-Nx is the output length of the Extract function, as specified in {{dependencies}}.
+Input:
+- client_secret, the ephemeral secret key of the Diffie-Hellman key exchange
+- client_identity,  the optional encoded client identity, which is set to its public key if not specified
+- client_private_key, the client's private key
+- server_identity, the optional encoded server identity, which is set to its public key if not specified
+- server_public_key, the server's public key
+- ke1, a KE1 message structure
+- ke1, a KE2 message structure
 
-The Derive-Secret parameter `preamble` is computed as:
+Output:
+- ke3, a KE3 structure
+- server_info, the server_info sent by the server
+- session_key, the shared session secret
 
+Steps:
+1. ikm = ClientIKM(client_secret, client_private_key, KE2.server_keyshare, server_public_key)
+2. preamble = Preamble(client_identity, KE1, server_identity, KE2.inner_ke2)
+3. Km2, Km3, handshake_encrypt_key, session_key = DeriveKeys(ikm, preamble)
+4. expected_server_mac = MAC(Km2, Hash(concat(preamble, KE2.enc_server_info))
+5. If !ct_equal(KE2.server_mac, expected_server_mac),
+     raise MacError
+6. client_mac = MAC(Km3, Hash(concat(preamble, KE2.enc_server_info, expected_server_mac))
+7. pad = Expand(handshake_encrypt_key, "EncryptionPad", len(KE2.enc_server_info))
+8. server_info = xor(pad, enc_server_info)
+9. session_key = DeriveSecret(prk, "session secret", preamble)
+10. Create KE3 ke3 with client_mac
+11. Output (ke3, server_info, session_key)
 ~~~
-preamble = concat("3DH",
-                  I2OSP(len(client_identity), 2), client_identity,
-                  KE1,
-                  I2OSP(len(server_identity), 2), server_identity,
-                  KE2.inner_ke2)
+
+~~~
+AKEServerFinalize(expected_client_mac, ke3)
+
+Input:
+- expected_client_mac, the mac over the complete transcript pre-computed in `AKEResponse()`
+- ke3, a KE3 structure
+
+Output:
+- bool, the boolean indicating whether the mac in KE3 is valid
+
+Steps:
+1. Output !ct_equal(KE3.client_mac, expected_client_mac)
+~~~
+
+###### Helper functions {#opaque-3dh-helper-functions)
+
+Random AKE Secret Key Share Generation Function:
+- RandomPrivateKey(): Generate a new scalar in the field of the AKE group using a byte string of length `Nsk` bytes
+  sourced from a cryptographically secure pseudorandom number generator (CSPRNG).
+
+~~~
+ServerIKM(server_secret, server_private_key, client_keyshare, client_public_key)
```

Comment:
Can we use the same function for both client and server here? In my implementation, that's what I do.